### ISSUE TITLE
Make DatabaseConfig a Provider or Inject it

### ISSUE THREAD

**Issue by schmitch** [playframework/playframework#5249]
Hello, currently we can only configure the database via the application.conf, however in some cases this is hard. 
Example:
if you use Environmentvariables to configure your Database you could get into trouble, especially if you don't follow the Heroku format of DATABASE_URL, the fix by itself would be really really simple.

Just make https://github.com/playframework/playframework/blob/702e89841fc54f5603a0d981c3488ed9883561fe/framework/src/play-jdbc/src/main/scala/play/api/db/DatabaseConfig.scala#L19 a class (Singleton) which could be injected and then inject it into the Java/Scala Config Pool.

**Comment by schmitch** [comment]
I closed the PR since I thought it is easier than it looked at first.
If somebody could gave me a hint how it would be the prefered way I will try it again.

The problem mostly comes through the static Injections Components.

**Comment by richdougherty** [comment]
I had a look at the DB code and Play doesn't seem to have simple hooks for overriding config at a per-database level. What you need to do is override the `DBModule` which class. This is the class that is responsible for reading the config and creating the the `DbApi` binding and a `Database` binding for each configured database. It's at this point that you have full control over config and how it is translated into actual database connections.

You can start by copying the `DbModule` source into your project then adapting it to suit your needs. Once you've written your own custom `DbModule` you'll want to enable your custom one and disable the default one.

```
play.modules.enabled += "modules.MyDbModule"
play.modules.disabled += "play.api.db.DbModule"
```

**Comment by schmitch** [comment]
Yeah for my case it would work, but It would be great if I could do something for everybody.

Also I looked at it too, and I would need to change a lot of things and if there would be some Provider-ish API it would help everybody in changing it.

**Comment by richdougherty** [comment]
In some projects, e.g. Akka, you can say something like `type = com.example.MyDatabaseConfigurator`. The "configurator" is created through through reflection and then it is passed the configuration and asked to create a `Database`. This lets you interpret the configuration in any way you want. You can even ignore the configuration if you like.

I actually have a nice general "configurator" system that I hope to integrate into Play. It reads configuration and then creates bindings. It is very flexible and supports things like conversion between Scala and Java APIs. One of the main use cases in Play would be database configuration. Unfortunately I'm a bit busy right now and don't have time to do it!
